infrastructure/src/test/java/org/corfudb/infrastructure/CorfuLayoutServerTest.java

```diff
@@ -0,0 +1,710 @@
+     * @param responseHeader  the header from the response message
+     * @return                true if the two headers have the same base field values
+     */
+    private boolean compareBaseHeaderFields(HeaderMsg requestHeader, HeaderMsg responseHeader) {
+        return requestHeader.getRequestId() == responseHeader.getRequestId() &&
+                requestHeader.getEpoch() == responseHeader.getEpoch() &&
+                requestHeader.getClientId().equals(responseHeader.getClientId()) &&
+                requestHeader.getClusterId().equals(responseHeader.getClusterId());
+    }
+
+    /**
+     * Perform the required preparation before running individual tests.
+     * This includes preparing the mocks and initializing the DirectExecutorService.
+     */
+    @Before
+    public void setup() {
+        mockServerContext = mock(ServerContext.class);
+        mockServerRouter = mock(IServerRouter.class);
+        mockChannelHandlerContext = mock(ChannelHandlerContext.class);
+        mockDataStore = mock(DataStore.class);
+
+        // Initialize with newDirectExecutorService to execute the server RPC
+        // handler methods on the calling thread
+        when(mockServerContext.getExecutorService(anyInt(), anyString()))
+                .thenReturn(MoreExecutors.newDirectExecutorService());
+
+        // Initialize an in memory DataStore for LayoutServer
+//        when(mockServerContext.getDataStore()).thenReturn(
+//                new DataStore(new ImmutableMap.Builder<String, Object>()
+//                .put("--memory", true)
+//                .build(), fn -> { }));
+        when(mockServerContext.getDataStore()).thenReturn(mockDataStore);
+
+        layoutServer = new LayoutServer(mockServerContext);
+    }
+
+    /**
+     * Test that the LayoutServer correctly handles a BOOTSTRAP_LAYOUT_REQUEST.
+     */
+    @Test
+    public void testBootstrapLayoutAck() throws IOException {
+        Layout l = getDefaultLayout();
+        RequestMsg request = getRequestMsg(
+                getBasicHeader(true, true),
+                getBootstrapLayoutRequestMsg(l)
+        );
+
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        layoutServer.handleMessage(request, mockChannelHandlerContext, mockServerRouter);
+        verify(mockServerRouter).sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+
+        ResponseMsg response = responseCaptor.getValue();
+        // Assert that the payload has a BOOTSTRAP_LAYOUT_RESPONSE and that the base
+        // header fields have remained the same
+        assertTrue(compareBaseHeaderFields(request.getHeader(), response.getHeader()));
+        assertTrue(response.getPayload().hasBootstrapLayoutResponse());
+        assertTrue(response.getPayload().getBootstrapLayoutResponse().getBootstrapped());
+    }
+
+    /**
+     * Test that the LayoutServer correctly handles a BOOTSTRAP_LAYOUT_REQUEST with null clusterId.
```

Comment:
Maybe consider adding a cause of a null clusterId in the comment description (i.e. ... correctly handles a BOOTSTRAP_LAYOUT_REQUEST with a legacy layout having a null clusterId)